Adding Planning Context Plugin Framework to OMPL Planning Plugin

zkingston opened the pull request:
### Description

Fixes https://github.com/ros-planning/moveit/issues/117 by adding a plugin framework to easily incorporate adding new OMPL-based planning contexts without having to recompile the moveit_ompl_planners plugin.

The new interface works with the existing `ompl_planning.yaml` files that are output with the _MoveIt_ setup wizard.  You can simply extend the `.yaml` planner specification to specify the `plugin` where the `ompl_interface::PlannerContextManager` can create the planning context for the given planner.

Here is an example `ompl_planning.yaml` that points out the new feature:
```yaml
planner_configs:
  RRTkConfigDefault:
    plugin: ompl_interface/ModelBasedPlanningContext  # This line was added. This is the fully-qualified name of the plugin. ModelBasedPlanningContext is the default option.
    type: geometric::RRT
    range: 0.0
    goal_bias: 0.05
  RRTConnectPose:
    plugin: ompl_interface/PoseModelBasedPlanningContext  # This line was added. Here, we specify a planning context plugin that uses the PoseModelBasedStateSpace always.
    type: geometric::RRTConnect
    range: 0.0
  RRTConnectkConfigDefault: # Note that no line needs to be added, and the ModelBasedPlanningContext will be used by default.
    type: geometric::RRTConnect
    range: 0.0
```
Additionally, the setup assistant has been modified to output the default plugin in the configuration file (`plugin: ompl_interface/ModelBasedPlanningContext`).

The interface is plugin centric, where the planning context is responsible for the majority of its own configuration.  New plugins must derive from `ompl_interface::OMPLPlanningContext`.  For standard geometric planning, it is possible to derive from the existing `ompl_interface::ModelBasedPlanningContext` class and simply configure your new planner. By default, there are three plugins provided:

1. `ompl_interface::ModelBasedPlanningContext`: The default planning context that was used before adding the plugin framework.
2. `ompl_interface::JointModelBasedPlanningContext`: A planning context that is `ompl_interface::ModelBasedPlanningContext` but will always choose `JointModelBasedStateSpace`.
3. `ompl_interface::PoseModelBasedPlanningContext`: A planning context that is `ompl_interface::ModelBasedPlanningContext` but will always choose `PoseModelBasedStateSpace`.

Note that the addition of plugins 2 & 3 removes the need for https://github.com/ros-planning/moveit/pull/541 (whose features are still implemented in plugin 1), and is left to not break existing API.

Additionally, the `ModelBasedStateSpace` has been updated to inherit from OMPL's `RealVectorStateSpace`. This enables a few of the new planners in OMPL to work with _MoveIt_, such as BIT* and Informed RRT*, which are both added to `ModelBasedPlanningContext` and to the setup assistant.

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)

zkingston commented:
I forgot, another thing to note is that the `ModelBasedStateSpace` now derives from `RealVectorStateSpace` so that BIT* works.

davetcoleman commented:
Skimmed this huge PR... excited! Can we expect to see this PR to the ros-planning/moveit soon?

mamoll commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/parameterization/model_based_state_space.h:
```diff
@@ -179,26 +179,27 @@ class ModelBasedStateSpace : public ompl::base::StateSpace
 
-  virtual ompl::base::State* allocState() const;
-  virtual void freeState(ompl::base::State* state) const;
+  virtual ob::State* allocState() const;
```
Two general style comments:
- I prefer full namespace names: `ompl::base` instead of `ob`
- I prefer `override` over `virtual` to different methods that are marked overridable from ones that doing the overriding.

mamoll commented on moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp:
```diff
@@ -53,51 +55,177 @@
+
+void ompl_interface::ModelBasedPlanningContext::registerDefaultPlanners()
+{
+  registerPlannerAllocator("geometric::RRT", boost::bind(&allocatePlanner<og::RRT>, _1, _2, _3));
```
lambda expressions are generally preferable (for legibility and performance) over {boost,std}::bind.

zkingston commented:
@davetcoleman you should see it soon (within the week, hopefully)! I wanted to get some preliminary feedback from members of the lab who use the old Luna plugin before submitting it to ros-planning/moveit.

BryceStevenWilley commented on moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp:
```diff
@@ -53,51 +55,177 @@
+    ompl_simple_setup_->setup();
+}
+
+namespace
```
Any reason for the anonymous namespace? It's named in the original plugin, and from what I can tell, anonymous namespaces are really only necessary for objects/structs.

BryceStevenWilley commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/ompl_planning_context.h:
```diff
@@ -0,0 +1,335 @@
+  }
+
+  /**
+   * \brief Sets whether or not the context is set to simplify solutions.
```
Should be 'sets whether or not the context *will* simplify solutions'?

mamoll commented on moveit_planners/ompl/ompl_interface/src/model_based_planning_context.cpp:
```diff
@@ -53,51 +55,177 @@
+#include <ompl/geometric/planners/rrt/TRRT.h>
+#include <ompl/geometric/planners/rrt/BiTRRT.h>
+#include <ompl/geometric/planners/rrt/LBTRRT.h>
+
```
... or really any of the other planners in the ompl::geometric namespace that don't require any special setup. See <http://ompl.kavrakilab.org/classompl_1_1base_1_1Planner.html>

zkingston commented on moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/ompl_planning_context.h:
```diff
@@ -0,0 +1,335 @@
+   */
+  void setCompleteInitialState(const robot_state::RobotState& complete_initial_robot_state)
+  {
+    *complete_initial_robot_state_ = complete_initial_robot_state;
```
`complete_initial_robot_state_` is a member variable, and `complete_initial_robot_state` is passed in. On 122, `=` calls the copy constructor that copies `complete_initial_robot_state` into  `complete_initial_robot_state_`, and then  `complete_initial_robot_state_` is updated.

Note the trailing `_` on one of those.

zkingston commented:
Just addressed some of the comments @mamoll and @BryceStevenWilley left. `override` was used where applicable, `csm_` was changed back to `constraint_sampler_manager_`, full OMPL namespaces were used rather than the shortened versions, and all planners that are available in OMPL are now provided in `ModelBasedPlanningContext`, including I-RRT*.

I'll address a few of the {boost,std}::bind uses in the code. Also, I'm not sure about the license, I'll just leave it for now and see what they have to say at the mainline repo.

mamoll commented:
:+1: This looks like it must have been a ton of work. Nice job. I am sorry I can't give you more substantive comments.

BryceStevenWilley commented:
Looks good. Feature wise, this definitely looks like the minimum amount that we'd want to merge.  And the copyright can stay I guess. Those were files you wrote yourself though, right?

zkingston commented:
True, I just updated the dates in the copyrights.

zkingston commented:
@KeliangH done.

Thanks for the comments everyone, I'll get this ready for the PR to the upstream repo soon.
